File: apps/macos/VoxylioKit/Sources/VoxylioKit/Grouping.swift
// Sentence reconstruction for progressive captions — port of
// packages/core/src/grouping.js. Invariants (identical to JS):
//  - a group's `id` NEVER depends on its mutable text — only on its start;
//  - `version` changes whenever the text changes;
//  - the LAST group of a live stream is a draft (final == false);
//  - roll-up AND sliding-window feeds are merged into one cue.

import Foundation

public let groupMaxLen = 280 // max characters per sentence (safety cap)
public let groupMaxGap = 1.4 // silence (s) that closes a sentence

/// Cheap stable content hash (FNV-1a, 32-bit over UTF-16 units, base36) —
/// bit-identical to the JS `textHash`.
public func textHash(_ s: String) -> String {
    var h: UInt32 = 0x811C_9DC5
    for unit in s.utf16 {
        h ^= UInt32(unit)
        h = h &* 0x0100_0193
    }
    return String(h, radix: 36)
}

private let wordPunct: Set<Character> = [
    ".", ",", "!", "?", "…", ";", ":", "'", "\"", "«", "»", "(", ")", "[", "]",
]

func normalizeWords(_ s: String) -> [String] {
    let mapped = String(s.lowercased().map { wordPunct.contains($0) ? " " : $0 })
    return mapped.split(whereSeparator: { $0.isWhitespace }).map(String.init)
}

/// Longest overlap (in words) between the END of `a` and the START of `b`.
public func wordOverlap(_ a: String, _ b: String, minWords: Int = 2) -> Int {
    let aw = normalizeWords(a)
    let bw = normalizeWords(b)
    let maxN = min(aw.count, bw.count)
    guard maxN >= minWords else { return 0 }
    for n in stride(from: maxN, through: minWords, by: -1) {
        var match = true
        for i in 0..<n where aw[aw.count - n + i] != bw[i] {
            match = false
            break
        }
        if match { return n }
    }
    return 0
}

public struct RollupMerge: Equatable {
    public let text: String
    public let end: Double
    public let grew: Bool
    public init(text: String, end: Double, grew: Bool) {
        self.text = text
        self.end = end
        self.grew = grew
    }
}

/// Merge an incoming caption into the previous cue when the feed is
/// progressive (roll-up growth or sliding window). Returns nil when the
/// incoming cue starts a new sentence.
public func mergeRollup(
    lastText: String?, lastEnd: Double, start: Double, end: Double, text: String
) -> RollupMerge? {
    guard let last = lastText else { return nil }
    if start > lastEnd + 0.6 { return nil }

    // Prefix relation (classic roll-up)
    if text.hasPrefix(last) || last.hasPrefix(text) {
        return RollupMerge(
            text: text.count > last.count ? text : last,
            end: max(lastEnd, end),
            grew: text.count > last.count
        )
    }

    // Sliding window: longest suffix of last == prefix of incoming
    let overlap = wordOverlap(last, text, minWords: 2)
    if overlap > 0 {
        let bw = normalizeWords(text)
        if overlap >= bw.count {
            // incoming is entirely contained in the tail of last
            return RollupMerge(text: last, end: max(lastEnd, end), grew: false)
        }
        // Replace last's overlapping tail with the FULL incoming text (the
        // incoming version of the shared words carries the richest
        // punctuation). Token starts mirror the JS /\S+/g scan.
        var starts: [String.Index] = []
        var inWord = false
        var i = last.startIndex
        while i < last.endIndex {
            if !last[i].isWhitespace {
                if !inWord {
                    starts.append(i)
                    inWord = true
                }
            } else {
                inWord = false
            }
            i = last.index(after: i)
        }
        let cutIdx = starts.count >= overlap ? starts[starts.count - overlap] : last.startIndex
        let head = String(last[last.startIndex..<cutIdx])
            .replacingOccurrences(of: "\\s+$", with: "", options: .regularExpression)
        let merged = head.isEmpty ? text : head + " " + text
        return RollupMerge(
            text: merged,
            end: max(lastEnd, end),
            grew: merged.count > last.count
        )
    }
    return nil
}

public struct Cue: Equatable {
    public let start: Double
    public let end: Double
    public let text: String
    public init(start: Double, end: Double, text: String) {
        self.start = start
        self.end = end
        self.text = text
    }
}

public struct Group: Equatable {
    public var start: Double
    public var end: Double
    public var text: String
    public var id: String
    public var version: String
    public var final: Bool
    /// Display/debug only — NEVER use as identity.
    public var key: String
}

/// Rebuild sentence groups from cues — deterministic, identical to JS.
public func buildGroups(_ cues: [Cue], maxLen: Int = groupMaxLen, maxGap: Double = groupMaxGap) -> [Group] {
    struct Building {
        var start: Double
        var end: Double
        var text: String
    }
    var groups: [Building] = []
    var cur: Building?
    for c in cues {
        let txt = cleanCaption(c.text)
        if txt.isEmpty { continue }
        if let current = cur,
            endsSentence(current.text) || c.start - current.end > maxGap
            || current.text.count > maxLen
        {
            groups.append(current)
            cur = nil
        }
        if cur == nil {
            cur = Building(start: c.start, end: c.end, text: txt)
        } else if cur!.text.hasSuffix(txt) {
            // Duplicated fragment (progressive captions): extend, don't repeat
            cur!.end = max(cur!.end, c.end)
        } else {
            cur!.end = max(cur!.end, c.end)
            cur!.text += " " + txt
        }
    }
    if let current = cur { groups.append(current) }

    return groups.enumerated().map { i, g in
        let id = "g" + String(Int((g.start * 100).rounded()))
        let version = textHash(g.text)
        return Group(
            start: g.start,
            end: g.end,
            text: g.text,
            id: id,
            version: version,
            final: i < groups.count - 1,
            key: id + ":" + version
        )
    }
}

File: packages/core/src/grouping.js
// Sentence reconstruction for progressive captions.
//
// Key invariants (see docs/IMPLEMENTATION_PLAN.md and the repetition-bug
// post-mortem in the tests):
//  - a group's `id` NEVER depends on its mutable text — only on its start;
//  - `version` changes whenever the text changes, so consumers can detect
//    growth without re-keying;
//  - the LAST group of a live stream is a `draft` (still growing) — only
//    the caller can finalize it (time-stability heuristics live there);
//  - roll-up AND sliding-window caption feeds are merged into one cue.
import { cleanCaption, endsSentence, continuesEllipsis } from "./subtitles.js";

export const GROUP_MAX_LEN = 280; // max characters per sentence (safety cap)
export const GROUP_MAX_GAP = 1.4; // silence (s) that closes a sentence

// Cheap stable content hash (FNV-1a, 32-bit) for group versions.
// Math.imul keeps the multiply in true 32-bit integer arithmetic — the
// previous float multiply silently rounded low bits above 2^53, which
// no integer-exact port could ever reproduce. Hashes are only compared
// within a session, so changing the algorithm was safe.
export function textHash(s) {
  let h = 0x811c9dc5;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 0x01000193) >>> 0;
  }
  return h.toString(36);
}

function normalizeWords(s) {
  return s
    .toLowerCase()
    .replace(/[.,!?…;:'"«»()\[\]]/g, " ")
    .split(/\s+/)
    .filter(Boolean);
}

// Longest overlap (in words) between the END of `a` and the START of `b`.
// Returns the number of overlapping words of `b`, or 0.
export function wordOverlap(a, b, minWords = 2) {
  const aw = normalizeWords(a);
  const bw = normalizeWords(b);
  const max = Math.min(aw.length, bw.length);
  for (let n = max; n >= minWords; n--) {
    let match = true;
    for (let i = 0; i < n; i++) {
      if (aw[aw.length - n + i] !== bw[i]) {
        match = false;
        break;
      }
    }
    if (match) return n;
  }
  return 0;
}

// Merge an incoming caption into the previous cue when the feed is
// progressive. Handles BOTH shapes real players produce:
//  - roll-up:        "Welcome" → "Welcome to the course"      (prefix growth)
//  - sliding window: "Welcome to the course" → "to the course of coding"
//    (the window drops leading words; the overlap must be stitched).
// Returns { text, end, grew } when it merges, or null.
export function mergeRollup(last, start, end, text) {
  if (!last) return null;
  if (start > last.end + 0.6) return null;
  // A cue must not grow forever: an uninterrupted roll-up monologue would
  // otherwise become one giant, permanently-draft group (never spoken as
  // long as it keeps changing, then voiced as a monster utterance). Close
  // it at a sentence boundary once well past the group cap, and refuse
  // outright at a hard ceiling even mid-sentence.
  if (
    (last.text.length > GROUP_MAX_LEN * 3 && endsSentence(last.text)) ||
    last.text.length > 2000
  ) {
    return null;
  }

  // Prefix relation (classic roll-up)
  if (text.startsWith(last.text) || last.text.startsWith(text)) {
    return {
      text: text.length > last.text.length ? text : last.text,
      end: Math.max(last.end, end),
      grew: text.length > last.text.length,
    };
  }

  // Sliding window: longest suffix of last == prefix of incoming
  const overlap = wordOverlap(last.text, text, 2);
  if (overlap > 0) {
    const bw = normalizeWords(text);
    if (overlap >= bw.length) {
      // incoming is entirely contained in the tail of last
      return { text: last.text, end: Math.max(last.end, end), grew: false };
    }
    // Replace last's overlapping tail with the FULL incoming text: the
    // incoming version of the shared words carries the richest
    // punctuation ("…de codage." vs "…de codage").
    // The overlap is counted in NORMALIZED words (apostrophes and
    // ellipses split: "don't" = 2 words) but the cut walks \S+ tokens
    // ("don't" = 1 token) — so consume tokens from the end until they
    // account for `overlap` normalized words, instead of cutting
    // `overlap` tokens (which deleted extra words: "I said don't stop" +
    // "don't stop now" once lost "said").
    const re = /\S+/g;
    const starts = [];
    const tokens = [];
    let m;
    while ((m = re.exec(last.text)) !== null) {
      starts.push(m.index);
      tokens.push(m[0]);
    }
    let cutIdx = 0;
    let consumed = 0;
    for (let i = tokens.length - 1; i >= 0; i--) {
      consumed += normalizeWords(tokens[i]).length;
      if (consumed >= overlap) {
        cutIdx = starts[i];
        break;
      }
    }
    const head = last.text.slice(0, cutIdx).trimEnd();
    const merged = head ? head + " " + text : text;
    return {
      text: merged,
      end: Math.max(last.end, end),
      grew: merged.length > last.text.length,
    };
  }
  return null;
}

// Rebuild sentence groups from cues. The result is deterministic for a
// given cue list; each group carries:
//   id      — stable, start-derived, text-independent
//   version — text hash, changes when the group grows
//   final   — false ONLY for the trailing group (it may still grow);
//             time-based stabilization is the caller's job.
export function buildGroups(cues, opts = {}) {
  const MAX_LEN = opts.maxLen ?? GROUP_MAX_LEN;
  const MAX_GAP = opts.maxGap ?? GROUP_MAX_GAP;
  const groups = [];
  let cur = null;
  for (const c of cues) {
    const txt = cleanCaption(c.text);
    // Letterless leftovers ("...", "???", "1:23") are display junk: they
    // must never reach a translator, a quota meter, or a voice.
    if (!txt || !/\p{L}/u.test(txt)) continue;
    if (
      cur &&
      ((endsSentence(cur.text) && !continuesEllipsis(cur.text, txt)) ||
        c.start - cur.end > MAX_GAP ||
        cur.text.length > MAX_LEN)
    ) {
      groups.push(cur);
      cur = null;
    }
    if (!cur) {
      cur = { start: c.start, end: c.end, text: txt };
    } else if (cur.text.endsWith(txt)) {
      // Duplicated fragment (progressive captions): extend, don't repeat
      cur.end = Math.max(cur.end, c.end);
    } else {
      cur.end = Math.max(cur.end, c.end);
      cur.text += " " + txt;
    }
  }
  if (cur) groups.push(cur);
  // Same-start groups (two speakers on one timestamp, frame-quantised DOM
  // cues) must NOT share an id — a collision silently swallows the second
  // line (marked spoken with the first, translation dropped as stale).
  // The disambiguating counter is deterministic for a given cue list, so
  // ids stay stable across rebuilds.
  const used = new Map();
  for (let i = 0; i < groups.length; i++) {
    const g = groups[i];
    let id = "g" + Math.round(g.start * 100);
    const n = used.get(id) || 0;
    used.set(id, n + 1);
    if (n > 0) id += "_" + n;
    g.id = id;
    g.version = textHash(g.text);
    // The trailing group may still be growing (live/progressive feeds).
    g.final = i < groups.length - 1;
    // Kept for display/debug only — NEVER use as identity.
    g.key = g.id + ":" + g.version;
  }
  return groups;
}
